### DISCUSSION TITLE
What files do I have to modify to program a new control law?

### DISCUSSION THREAD

**Discussion by Carmol1** [bitcraze/discussions#64] [Q&A]
Hello.

I am working with the crazyswarm package since I want to control a group of crazyflie 2.1, for this reason, my questions are these, what files should I modify to program my control law? and how can I make use of the information obtained of the optitrack cameras?.

I'd appreciate your help.

**Comment by krichardsson** [comment]
Hi!

> what files should I modify to program my control law?

Could you please specify a bit more in detail what behavior you want to change? "control law" can be interpreted in may ways and there are control algorithms on multiple levels.

> and how can I make use of the information obtained of the optitrack cameras?.

Position information from external positioning systems are sent to the Crazyflie through the localization service via the Crazyradio. In the Crazyflie you can find the localization service here https://github.com/bitcraze/crazyflie-firmware/blob/2022.05/src/modules/src/crtp_localization_service.c
The information is passed on to active estimator through the https://github.com/bitcraze/crazyflie-firmware/blob/2022.05/src/modules/interface/estimator.h#L97 method.
Usually you just pass the positioning information from your mocap system on to the Crazyflie, either using Crazyswarm or the python lib. In the Crazyflie the information is fed into the estimator which enables it to estimate its current position, which in turn is used by the controller to control the motors.

**Reply by Carmol1** [reply]
Hello. 

Thank you very much for answering, regarding the control law, I want to perform trajectory tracking, so I have designed a control law for the position part and another for the orientation, but I don't know in which file I have to make changes.

On the other hand, I thank you for the information provided in the second question, I will check it.

Greetings.

**Reply by krichardsson** [reply]
Sorry for the late reply!

This sounds very much like the functionality in the high level commander, it is continuously generating set points (when enabled) that makes the Crazyflie follow a trajectory. The set points are fed into the controller that adjusts the motor speeds to move from the current state to the desired state of the set point.

The high level commander is mainly implemented in 
* https://github.com/bitcraze/crazyflie-firmware/blob/master/src/modules/src/crtp_commander_high_level.c
* https://github.com/bitcraze/crazyflie-firmware/blob/master/src/modules/src/planner.c

Maybe this can give you some ideas?

You can also read a bit more about how the system works, all the way from sensor data to controlling the platform https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/functional-areas/sensor-to-control/

**Reply by Carmol1** [reply]
Hi.

No worries, I thank you for your reply, however, I still have many doubts about how implement my controller, for the moment I have been revising the Mellinger controller to understand than you did, but it is hard to figure out all the program, at least for me, for such motive, I would like to know, the following 

Do these instruction setpoint->position.x, setpoint->position.y, setpoint->position.z store the informatiom of the optitrack cameras or other positioning system to be used in the  controller?.

Is it always necessary to use this part 

void controllerMellingerInit(void)
{
  controllerMellingerReset();
}

bool controllerMellingerTest(void)
{
  return true;
}

void controllerMellinger(control_t *control, setpoint_t *setpoint,
                                         const sensorData_t *sensors,
                                         const state_t *state,
                                         const uint32_t tick)

to use the instruction mentioned in the first question?.


What function do these part have?.

setpoint->mode.x == modeAbs

setpoint->mode.yaw == modeVelocity

Finally, I would like to know, how can I conbine the information of position among crazyflies in the controller? .

For example to use this
![SmartSelect_20220615-005216_Drive](https://user-images.githubusercontent.com/104467412/173775656-eff3e3c1-4065-49a7-b66f-1c0eec803965.jpg)
in the transtaltion control law, where x_i is the position of first crazyflie, x_j the position of the crazyflie neighbor and v_i and v_j are respectively the velocity of the first crazyflie and its neighbor.

I'd appreciate your help.

**Reply by krichardsson** [reply]
> Do these instruction setpoint->position.x, setpoint->position.y, setpoint->position.z store the informatiom of the optitrack cameras or other positioning system to be used in the controller?

No, the setpoint is the desired position (state) that you want the Crazyflie to have, [this is the goal of the controller](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/functional-areas/sensor-to-control/controllers/#overview-of-control). 
The information from you mocap system should be fed to the estimator as any other sensor data to improve the estimated state ([see the basis structure of sensor to control](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/functional-areas/sensor-to-control/))

> Is it always necessary to use this part
void controllerMellingerInit(void)
{
controllerMellingerReset();
}
...

Please see description in #91

> What function do these part have?.
setpoint->mode.x == modeAbs
setpoint->mode.yaw == modeVelocity

[Please see](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/functional-areas/sensor-to-control/commanders_setpoints/)

I think a good starting point for you is to go back to [the link I sent you previously about sensor to control](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/functional-areas/sensor-to-control/)

**Reply by Carmol1** [reply]
Hi.

Thanks for your help. I am going to check the information, Finally, is it possible to conbine the information of position among crazyflies to use this part?
![SmartSelect_20220615-005216_Drive](https://user-images.githubusercontent.com/104467412/173775656-eff3e3c1-4065-49a7-b66f-1c0eec803965.jpg)

I'd appreciate your help.

**Reply by krichardsson** [reply]
I'm not exactly sure what  you mean. If you are interested in the position of multiple Crazyflies, this must be handled on a completely different level.

**Reply by Carmol1** [reply]
Hi.

Thanks for your prompt reply. I'm interested in to carry out the formation behavior, for this reason I need to program that part, but I'm not sure if that part can be programmed onboard.

Regards.

**Reply by krichardsson** [reply]
No, a Crazyflie does not have any information about the position of other Crazyflies

**Reply by Carmol1** [reply]
Hi.

Thanks for your reply, so, in this case is necessary to use an offboard control?.

Regards.

**Reply by krichardsson** [reply]
Yes

**Reply by Carmol1** [reply]
Hi.

Thanks for your help. So, I am going to begin to research information about the topic.

Regards.

**Comment by knmcguire** [comment]
@Carmol1  Just some moderation here as well, as your discussion mentions two issues (we rather would like to have one per discussion).

Since crazyswarm uses a different library to obtain optitrack information than we do, you should post a discussion about optitrack over there: https://github.com/USC-ACTLab/crazyswarm/discussions

You can continue here to talk about the controller question as I think that would be suitable here.

**Reply by Carmol1** [reply]
Hello.

I'm sorry, I appreciate your answer, on the other hand, I have already posted on that forum three times and have never received a reply.

Greetings.

**Reply by knmcguire** [reply]
Ah I'm sorry to hear that! In that case, could you start a new discussion here with only about the positioning cameras question (with a good descriptive title) ? This will make it easier for us to handle. 

Also try to include a link to those discussions you mentioned that weren't answered, so that we have a context. I'll try to ping the maintainers of the optitrack package that crazyswarm uses.

**Reply by knmcguire** [reply]
oh wait, it seems that you got an answer already on your discussion: https://github.com/USC-ACTLab/crazyswarm/discussions/655#discussioncomment-2866805. So please continue it there, as those are the direct maintainers.

**Reply by Carmol1** [reply]
Hello. 

Yes, I have already received an answer, I hope to continue receiving help in the coming days.

Greetings.

**Comment by knmcguire** [comment]
Quoted from @krichardsson in https://github.com/orgs/bitcraze/discussions/91 which gives initial instructions to add a new controller. @Carmol1 let us know if there is anything unclear here. 

> Hi!
> 
> Controllers are plugins (sort of) and should be pretty straight forward to add.
> 
> Controllers are registered in https://github.com/bitcraze/crazyflie-firmware/blob/master/src/modules/src/controller.c#L24-L29 As you can see you have to implement 3 functions (init, test and update) that you pass on to the system in the registration. Add a new line in the registration after the INDI controller and add your functions here.
> 
> Add a new identifier in https://github.com/bitcraze/crazyflie-firmware/blob/master/src/modules/interface/controller.h#L31-L37 (after the ControllerTypeINDI but before ControllerType_COUNT)
> 
> Add a new c-file for your controller and add it to the Kbuild file https://github.com/bitcraze/crazyflie-firmware/blob/master/src/modules/src/Kbuild#L10-L13.
> 
> You can start from the mellinger controller if you like or just make a new file, the only thing that is important is that you implement the three functions.
> 
> Your init function should initialize data. It will be called when the controller is activated. Your test function should just return true.
> 
> Your update function will be called from the stabilizer loop (at 1000 Hz) and the task is essentially to fill in the values in the control struct (roll, pitch, yaw and thrust). You get at bunch of information in the other structs that you can use to understand the state of the world.
>

**Reply by knmcguire** [reply]
Also, perhaps an question we should have asked sooner: do you want to have this controller **onboard** or **offboard**? All the above instructions are for programming an new _onboard_ controller inside the firmware which is complicated, but if this is not what you meant you should probably mention that now 🙂 If you rather want to have the controller exist on the computer for instance.
